# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/raju-cse/Flyrank-internship_ml/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

## My Lane 1

I chose the **Google Search Ranking & Discoverability** lane because I want to understand which webpage characteristics are associated with better search performance.

My goal is to use the provided search data to identify pages that may need optimization and support better SEO decisions. This project focuses on helping prioritize content updates using data rather than intuition alone.

In [8]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

## Research Question

**Research Question:** Which webpage characteristics are associated with declining search performance, and how can they help prioritize content refresh?

**Decision:** Decide which webpages should be reviewed and refreshed first.

**Action:** SEO or content teams can prioritize updating pages that are most likely to benefit from a content refresh.

**Cost of a Wrong Recommendation:** If the recommendation is incorrect, valuable time and resources may be spent updating the wrong pages, while pages that actually need attention may continue to lose traffic.

**Why Machine Learning:** Machine learning can identify useful patterns across many webpage features and support data-driven decisions more effectively than manual inspection alone.

In [9]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

## Quick Look at the Data

The starter dataset contains **30,000 rows** of anonymized webpage data.

To better understand the dataset, I explored a few basic statistics such as the total number of rows, columns, and the distribution of trend directions. These initial observations show that the dataset is large enough to support meaningful analysis and model development during the internship.

In [10]:
import os

print(os.getcwd())
print(os.listdir())

/content/Flyrank-internship_ml
['LICENSE', 'notebooks', 'skills', '.github', 'CLAUDE.md', 'data', '.git', 'submission', 'AGENTS.md', 'docs', 'requirements.txt', 'outputs', 'README.md', 'work', 'DATA_USE.md', 'GUIDE.md', '.gitignore', 'SETUP.md', 'scripts']


In [11]:
import os
import subprocess

REPO_URL = "https://github.com/raju-cse/Flyrank-internship_ml"
REPO_DIR = "Flyrank-internship_ml"

if not os.path.exists(REPO_DIR):
    subprocess.run(["git", "clone", REPO_URL], check=True)

os.chdir(REPO_DIR)

print("Current directory:", os.getcwd())

Current directory: /content/Flyrank-internship_ml/Flyrank-internship_ml


In [12]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Rows:", len(df))
print("Columns:", len(df.columns))
print(df.head())

Rows: 30000
Columns: 44
             content_id          client_id  search_volume  competition  \
0  content_304f48230142  client_f369cb89fc           10.0         0.67   
1  content_a1fb4e703a9e  client_4e07408562           90.0         0.01   
2  content_9aa793d4d895  client_7f2253d7e2            0.0         0.00   
3  content_331d6c4de07b  client_19581e27de           10.0         0.00   
4  content_d99b7a2d90ca  client_3fdba35f04            0.0         0.00   

  competition_level   cpc     content_type    main_intent  word_count  \
0              HIGH  2.05  keyword article  transactional      3221.0   
1               LOW  0.05  keyword article  informational      2481.0   
2               LOW  0.00  keyword article  informational      3515.0   
3               LOW  0.00  keyword article     commercial         NaN   
4               LOW  0.00  keyword article  informational      2803.0   

   char_count  ... char_count_tier   ctr  avg_position  engagement_rate  \
0     20457.0  ..

In [13]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Rows:", len(df))
print("Columns:", len(df.columns))

print("\nTrend Direction Counts:")
print(df["trend_direction"].value_counts())

print("\nAverage CTR:", round(df["ctr"].mean(), 4))
print("Average Search Volume:", round(df["search_volume"].mean(), 2))

Rows: 30000
Columns: 44

Trend Direction Counts:
trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64

Average CTR: 0.5107
Average Search Volume: 158.88


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

This project aims to identify observed patterns in the available search data and provide decision-support for content optimization.

The results will be directional rather than definitive. They do not prove cause-and-effect relationships, and they do not predict Google's search algorithm. Any recommendations should be validated before making important business decisions.

## Quick Look at the Data (2–3 Real Numbers)

The starter dataset contains **30,000 rows** and **44 columns**, providing a substantial amount of anonymized search performance data for analysis.

Initial exploration shows that **16,262 pages are classified as "down"**, making declining page performance the largest trend category in the dataset. This suggests there are many opportunities to study why pages lose visibility and how they can be prioritized for content refresh.

The dataset also shows an **average CTR of 0.5107** and an **average search volume of 158.88**, indicating that it contains meaningful search performance metrics that can be used to support data-driven SEO decisions during this project.

In [14]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

## Careful Words



## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.